client/allocrunner/taskrunner/consul_hook.go

```diff
@@ -32,15 +32,17 @@ type consulHook struct {
 	task          *structs.Task
 	tokenDir      string
 	hookResources *cstructs.AllocHookResources
+	taskEnv       map[string]string
 	logger        log.Logger
 }
 
-func newConsulHook(logger log.Logger, tr *TaskRunner, hookResources *cstructs.AllocHookResources) *consulHook {
+func newConsulHook(logger log.Logger, tr *TaskRunner) *consulHook {
 	h := &consulHook{
 		task:          tr.Task(),
 		tokenDir:      tr.taskDir.SecretsDir,
-		hookResources: hookResources,
+		hookResources: tr.allocHookResources,
 	}
+	h.taskEnv = tr.envBuilder.Build().Map()
```

Comment:
The task env is a "builder", so when you call `Build` here you're finalizing it, rather than exposing the internal map to be written. You should make the builder part of the `consulHook` struct and then call a new `SetConsulToken` method on it (similar to the existing `SetVaultToken` method).

Reply:
fixed in [ad83aff](https://github.com/hashicorp/nomad/pull/19490/commits/ad83aff166ef2be0b4ceb617450fa4415280471f)